### ISSUE TITLE
Restrict allowed character sets for names in API (pipelines, jobs, resources, steps, etc)

### ISSUE THREAD

**Issue by vito** [concourse/concourse#3985]
## What challenge are you facing?

Tons of issues are opened all the time because various funny characters cause breakage in various places (e.g. `fly`, web UI):

* #2336
* #1892 
* #3941 
* #3787 
* #2211
* #3724
* ...probably like 10 others somewhere I can't find

We've never had any validation for these so folks have started to use all kinds of weird naming schemes which are difficult to support going forward, especially in contexts like `fly` where we often have our own notation in flags.

## What would make this better?

We should just decide on allowable characters and perform validation at pipeline submission time.

If I were to start Concourse over I would have restricted everything to `[a-z0-9-]+` and force everyone to use fairly simple job/pipeline names. I say `a-z` but I really mean alphabetical in whatever language the user speaks - I don't remember the regex notation for that.

I would even go as far as only allowing lowercase characters so we don't have to care about case-sensitivity, and allowing `-` but not `_` so we don't end up with ideological wars.

The advantage of having such a strict set of allowable characters would be that everyone's Concourse feels the same. This also allows us to use notation more freely, e.g. `/` for pipeline hierarchies (https://github.com/concourse/concourse/issues/532).

Are there any use cases out there for certain special characters in job/pipeline names?

How many people have I made angry with my recklessly narrow definition of allowed characters?

**Comment by pn-santos** [comment]
At my company we've already sort of have the convention around using `a-z0-9-` names for pipelines, tasks, jobs and resource names. Although sometimes you see `A-Z` and `_` here and there but it's the minority.

Given that most of these names end up in a URI it makes it much nicer if they are URI "safe" and follow standard best-practices around URI naming schemes (lowercase everywhere, `-` instead of `_` and avoid all characters that would require percent encoding)

**Comment by hfinucane** [comment]
We have a lot of long job names- "test cross-region VPC peering", "deploy to dev", that sort of thing. We could change them to work with no spaces and only dashes, but it might be annoying.

We have something else that's a little hairier- we have our own implementation of spaces using ERB, and for PRs we iterate over the task files in the 'ci' directory of the branch, and then create standalone jobs for each. Currently we're doing something like s/\//::/ to sanitize the names, but that would get less tenable if we were locked down to `a-zA-Z0-9-`.

**Comment by ukabu** [comment]
We have a naming convention for our team names and our pipeline are usually named after their source git repo. However no conventions for everything else. We do have names in languages other than english (french). I don't think we have really weird names. However, I don't really see why we should restrict which characters are allowed in names. Any UTF-8 (besides control characters) should be allowed IMO.
Maybe just exclude characters used in file systems like /, \, :, but really this should be more an explicit exclusion list rather that an inclusion list.

**Comment by marco-m** [comment]
I agree with vito proposal for the reasons exposed. To make only an example, I was very surprised when I saw people using `/` as part of a name, to me `/` is obviously a separator, without going to Unix, it is enough to think about URL syntax.
I am also very comfortable in making this a breaking change.

**Comment by robwhitby** [comment]
no spaces would be a big change for us, we tend to name jobs things like "deploy to dev" or "integration test", mainly so that it reads nicely in the interface.

**Comment by robwhitby** [comment]
Hi,

I'm not in favour of this change for a couple of reasons.

We have over 1000 pipelines that will have to be updated - which is a pain in itself, but after doing that a lot of jobs will lose their build history because the name has changed.

Also our users like to use human readable names for jobs and especially tasks. e.g. Deploy to Dev (onprem), Run Tests etc. etc. It helps make the pipeline understandable in the UI.

Maybe the restriction could be made a warning instead of an error? or the set of allowed characters expanded?

Thanks

**Comment by vito** [comment]
@robwhitby We have to bite the bullet at some point. :slightly_frowning_face: We made the mistake of allowing every character under the sun, and that has made all sorts of weird naming patterns spring up from team to team, let alone the practical challenges of designing notation around an arbitrary character set.

If tools like `go fmt` have shown us anything it's that the value of global consistency ultimately outweighs minor personal preferences. I don't really think that using spaces in names makes things any clearer. The abundance of programming languages and CLI tools that use camel-case, hyphens, underscores, etc. shows that for me. Using spaces just adds a little bit of friction everywhere - now I have to quote arguments in `fly`, which is annoying because `fly` is the primary way of interacting with Concourse.

Jobs can preserve their history if you configure [`old_name`](https://concourse-ci.org/jobs.html#job-old-name). We can ease the transition by providing a tool to automatically rename all of your jobs based on some common scheme. There's no reason we have to make the transition as painful as possible. :slightly_smiling_face:

**Comment by pn-santos** [comment]
Would it just make it unnecessary complex if there was a "displayName" type property? (which would fallback to `name` if undefined).

All references could use `name` (which would have the restricted character set) but the "displayName" value would be there for presentation (UI or fly outputs).

**Comment by agurney** [comment]
Some fresh numbers from our installation. We have 2397 pipelines, 17652 jobs, and 13977 resources. Of those, 320 pipelines, 3006 jobs, and 826 resources fail the regex. 

The main reasons are:
* Capital letters
* Underscores
* Spaces (almost all of which occur in job names, as with @robwhitby 's examples)
* Periods

There are a few other odd characters being used, like `&`, `!` and pairs of `(`...`)`, but not really a large number. None of those instances seem worth keeping. The `-` vs `_` thing has been personally frustrating to me many times in many contexts, so I'm very sympathetic to avoiding it for Concourse going forward. The capital letter cases are common, but don't seem too problematic to avoid.

Looking over the "offending" names, there are quite a lot of cases where these characters are justified in order to match some existing name: 
* domain names with `.`
* names of existing git repositories with `_`, not all of which we control
* version numbers with `.`, again often matching product versioning of products we don't control
 
I think it is very convenient to be able to have this match be possible. If I'm working with https://github.com/prometheus/client_golang (I'm not, just picking this from today's GH trending repos as an example) then I'd like to name things with `client_golang` rather than `client-golang` to _avoid_ a point of friction. Likewise for other external things where the name has already been chosen. In our case, just allowing `_` and `.` would make that possible, even if there is potential for confusion if people pick underscore as a word separator.

**Comment by simonjohansson** [comment]
I second @pn-santos 's suggestion. Allowing a display name would be great.

**Comment by vito** [comment]
@agurney Thanks for the detailed analysis, that really helps!

Mild :+1: to the idea of a display name; it seems fairly harmless, though I'm not sure how often I'd use such a thing myself. Some prior art in #3793 though it never went anywhere.

**Comment by skreddy6673** [comment]
Is there any way i can enforce Developers to use certain naming convention while creating team or pipeline or job?

I need a naming convention so that i can map the pipeline to the application it belongs. I know i can ask them to follow certain naming convention, but is there anyway i can force them to do?

Thanks,

**Comment by klhex** [comment]
> @agurney Thanks for the detailed analysis, that really helps!
> 
> Mild 👍 to the idea of a display name; it seems fairly harmless, though I'm not sure how often I'd use such a thing myself. Some prior art in #3793 though it never went anywhere.

@vito Well, now since the latest Concourse version is yelling deprecation warnings at people, I'm pretty sure the interest in a display will significantly increase. We're currently in the midst of renaming our pipelines/jobs because of the deprecation warnings and would be really happy about display names for improved (or rather re-improved) user experience.

**Comment by kallisti5** [comment]
I just ran into the *PAIN* this one inflicts.  Folks, please don't block underscores in pipeline/job names.
Architecture names are extremely common (build x86_64 packages, etc)

We template out 1000's of lines of yaml with the architecture... by dropping _'s you'll be breaking all of our pipelines.

**Comment by vito** [comment]
@kallisti5 Underscores are allowed again as of v6.7.3. :) So far that has definitely been the most significant friction - "x86_64" is a very good example.

Thanks again everyone for dealing with the warnings and giving feedback!

**Comment by kallisti5** [comment]
Thanks @vito !  It's definitely fair to cleanup the pipeline names, glad we were able to find a good balance.
